src/main/kotlin/frc/team5499/frc2018Kotlin/path/PathFollower.kt

```diff
@@ -0,0 +1,121 @@
+            var t2: Double = (-b + dis) / (2.0 * a)
+
+                if (t1 >= 0 && t1 <= 1) {
+                    var temp: Vector2 = d * t1
+                    lookahead = begin + temp
+                    break
+                } else if (t2 >= 0 && t2 <= 1) {
+                    var temp: Vector2 = d * t2
+                    lookahead = begin + temp
+                    break
+                }
+            }
+        }
+        @Suppress("MagicNumber")
+        if (lookahead == null) {
+            lookahead = mPath.getPoint(mPath.getPathLength() - 1)
+        } else {
+            var distanceBetweenRobotEnd: Double = Vector2.distanceBetween(
+                robotPos, mPath.getPoint(mPath.getPathLength() - 1))
+            if (distanceBetweenRobotEnd < Constants.LOOK_AHEAD_DISTANCE) {
+                lookahead = mPath.getPoint(mPath.getPathLength() - 1)
+            }
+        }
+        return lookahead
+    }
+
+    private fun calculateCurvature(
+        robotPos: Vector2,
+        lookAhead: Vector2,
+        robotAngle: Double,
+        angleToLookahead: Double
+    ): Double {
+        var a: Double = (1 / Math.tan(robotAngle))
+        var b: Double = -1.0
+        var c: Double = -(1 / Math.tan(robotAngle)) * robotPos.y + robotPos.x
+
+        var x: Double = Math.abs(a * lookAhead.y + b * lookAhead.x + c) / (Math.sqrt(a * a + b * b))
+        var curvature: Double = (2.0 * x) / (Math.pow(Constants.LOOK_AHEAD_DISTANCE, 2.0))
+
+        var side: Double = Math.signum(Math.sin(robotAngle) * (lookAhead.x - robotPos.x) -
+            Math.cos(robotAngle) * (lookAhead.y - robotPos.y))
+        @Suppress("MagicNumber")
+        return if (angleToLookahead <= 90.0) curvature * side else -curvature * side
+    }
+
+    fun update(robotPos: Vector2, newRobotAngle: Double): MutableList<Double> {
+        var output: MutableList<Double> = mutableListOf<Double>()
+        var robotAngle: Double = newRobotAngle
+        robotAngle = Math.toRadians(robotAngle)
+        if (robotAngle.equals(0)) { robotAngle = Constants.EPSILON }
+
+        var lookahead: Vector2 = calculateLookahead(robotPos, robotAngle)
+
+        var robotToLookahead: Vector2 = lookahead - robotPos
+
+        @Suppress("MagicNumber")
+        var robotDirection: Vector2 = Vector2.representHeadingWithUnitVector(
+            -Math.toDegrees(robotAngle) + 90)
+        var angleToLookahead: Double = Math.abs(Vector2.angleBetween(robotToLookahead, robotDirection))
+
+        var curvature: Double = calculateCurvature(robotPos, lookahead, robotAngle, angleToLookahead)
+
+        @Suppress("MagicNumber")
+        var averageVelocity: Double =
+        if (angleToLookahead <= 90.0)
+        mPath.getPointVelocity(mLastClosestPointIndex)
+        else -mPath.getPointVelocity(mLastClosestPointIndex)
+
+        output[0] = averageVelocity * (2.0 + (curvature * Constants.TRACK_WIDTH)) / 2.0
+        output[1] = averageVelocity * (2.0 - (curvature * Constants.TRACK_WIDTH)) / 2.0

```

Comment:
Use a `Pair<Double, Double>` for this instead of an array. its neater. you'll need to refactor in the auto command as well though